src/matcouply/penalties.py

```diff
@@ -1322,3 +1322,98 @@ def penalty(self, x):
         if not isinstance(x, list):
             raise TypeError("Cannot compute PARAFAC2 penalty of other types than a list of tensors")
         return 0
+
+
+class TemporalSmoothnessPenalty(MatricesPenalty):
+    r"""Impose temporal smoothness on the uncoupled factor matrices. Formally, this means adding 
+    the following penalty term:
+
+    .. math::
+
+    \lambda_B \sum_{k=2}^K \lVert \M{B}_k - \M{B}_{k-1} \rVert^2_F .
+
+    More information as well as experiments can be found in :cite:p:`chatzis2023timeaware`.
+
+    **The proximal operator**
+
+    Taking the partial derivative of the objective function with respect to each auxiliary
+    variable results in a tridiagonal system of equations. More specifically:
+
+    .. math::
+
+    \begin{flalign}
+        & \bigl( 2\lambda_{\M{B}} + \rho_{\M{B}_1}\bigr) \M{Z}_{\M{B}_1} -2\lambda_{\M{B}}\M{Z}_{\M{B}_2} = \rho_{\M{B}_1} \bigr( \M{B}_1 + \M{\mu}_{\M{B}_1} \bigl) \notag \\
+        & \bigl(4\lambda_{\M{B}}+\rho_{\M{B}_k}\bigr)\M{Z}_{\M{B}_k}-2\lambda_{\M{B}} \bigl(\M{Z}_{\M{B}_{k-1}}+\M{Z}_{\M{B}_{k+1}} \bigr) =\rho_{\M{B}_k} \bigr(\M{B}_k+\M{\mu}_{\M{B}_k}\bigl), \qquad k = 2,\ldots,K-1 \notag \\
+        & \vdots \notag \\
+        & \bigl( 2\lambda_{\M{B}} + \rho_{\M{B}_K}\bigr) \M{Z}_{\M{B}_K} -2\lambda_{\M{B}}\M{Z}_{\M{B}_{K-1}} = \rho_{\M{B}_K} \bigr( \M{B}_K + \,{\mu}_{\M{B}_K} \bigl) \notag
+    \end{flalign}
+
+    We then form a tridiagonal matrix :math:`\M{A}` and solve the system of equations using Thomas algorithm.
+
+    Keep in mind that once we apply a norm-based penalty in the factor matrices, we need to also penalize the
+    norm of all other factor matrices :cite:p:`roald2021parafac2` to overcome the scaling ambiguity of PARAFAC2.
+    This can be done, for example, by adding ridge penalties on the norms of the other two factor matrices.
+
+    Parameters
+    ----------
+    smoothness_l : float
+        Hyperparemeter controlling the strength of the smoothness penalty.
+    aux_init : {"random_uniform", "random_standard_normal", "zeros", tl.tensor(ndim=2), list of tl.tensor(ndim=2)}
+        Initialisation method for the auxiliary variables
+    dual_init : {"random_uniform", "random_standard_normal", "zeros", tl.tensor(ndim=2), list of tl.tensor(ndim=2)}
+        Initialisation method for the auxiliary variables
+    """
+
+    def __init__(self, smoothness_l, aux_init="random_uniform", dual_init="random_uniform"):
+        super().__init__(aux_init=aux_init, dual_init=dual_init)
+        self.smoothness_l = smoothness_l
+
+    @copy_ancestor_docstring
+    def factor_matrices_update(self, factor_matrices, feasibility_penalties, auxes):
+
+        B_is = factor_matrices
```

Comment:
```suggestion
        B_is = factor_matrices
        I = len(factor_matrices)  # noqa: E741
```

It might be more readible if we add a variable for the number of matrices.